File: Sorting/countingSort.py
"""
Counting Sort
Time Complexity - O(n+k), where n is the input size and k is the range of input(maximum element in the input)
Space Complexity - O(n+k)
"""

def counting(ls):
    auxiliary = [0] * (max(ls) + 1)

    for i in range(len(ls)):
        auxiliary[ls[i]] += 1

    sorted_list = []
    for i in range(len(auxiliary)):
        if (auxiliary[i] > 0):
            for j in range(auxiliary[i]):
                sorted_list.append(i)

    return sorted_list


#                                ### Testcases ###
#
# ls = [4,3,1,7,8,5,2,19,15,100,95,125,11]
# print (counting(ls))

File: Sorting/countingSort.cpp
/*
Counting sort - It is a variation of counting sort algorithm writtein in countingSort.py

Time Complexity - O(n)
Space Complexity - O(n) + O(k) 
*/

#include <iostream>
#include <vector>
using namespace std;

vector<int> countingSort(vector<int> &arr, vector<int> &result, int n, int k){
    vector<int> temporaryArr(k+1, 0);

    for(int i = 0; i < arr.size(); i++) temporaryArr[arr[i]] += 1;

    for(int i = 1; i <= k; i++) temporaryArr[i] += temporaryArr[i-1];

    for(int i = 0; i < n; i++){
        result[temporaryArr[arr[i]] - 1] = arr[i];
        temporaryArr[arr[i]] -= 1;
    }

    return result;
}

int main(){
    int n, k;
    cin>>n>>k;
    vector<int> arr(n);
    vector<int> result(n);

    for(int i = 0; i < n; i++)  cin>>arr[i];

    countingSort(arr, result, n, k);

    for(int i = 0; i < n; i++) cout<<result[i] << " ";
    
    return 0;
}